Backfill canceled subscription events

allouis opened the pull request:
refs https://github.com/TryGhost/Team/issues/1455

There are a few cases for canceled subscriptions:

1. They have now expired
2. They were canceled _before_ we added cancelation events
3. They were canceled _after_ we added cancelation events

For 1 we can ignore them, because the MRR has corrected itself.

For 2 we must insert an event with a negative mrr_delta
For 3 we must update the cancellation event with the negative mrr_delta

Because we calculate the mrr_delta from the mrr column of a subscription, this migration must run before the one where we update mrr based on cancelations: https://github.com/TryGhost/Ghost/commit/95baa87f96651003a97eac768a0596c8818307f0

allouis committed:
Backfilled canceled events for subscriptions

refs https://github.com/TryGhost/Team/issues/1455

allouis committed:
fixup! Backfilled canceled events for subscriptions

allouis commented on core/server/data/migrations/versions/5.0/2022-04-21-15-44-backfill-mrr-events-for-canceled-subscriptions.js:
```diff
@@ -0,0 +1,66 @@
+
+const {createTransactionalMigration} = require('../../utils');
+
+module.exports = createTransactionalMigration(
```
Will change this back to createIrreversibleMigrations

allouis commented on core/server/data/migrations/versions/5.0/2022-04-21-15-44-backfill-mrr-events-for-canceled-subscriptions.js:
```diff
@@ -0,0 +1,66 @@
+        }, {toUpdate: [], toInsert: []});
+
+        await knex.batchInsert('members_paid_subscription_events', toInsert);
+        for (const event of toUpdate) {
```
Not sure about this, I think there's probably some nested update query we can do - not sure if it's worth the complexity though

allouis commented on core/server/data/migrations/versions/5.0/2022-04-21-15-44-backfill-mrr-events-for-canceled-subscriptions.js:
```diff
@@ -0,0 +1,66 @@
+            logging.info(`Found ${subscriptions.length} canceled subscriptions, inserting MRR events.`);
+        }
+
+        const {toUpdate, toInsert} = subscriptions.reduce((memo, subscription) => {
```
We already loop here, so if performance is a concern, I can do the DB updates in this loop and generate the inserts to be batched

matthanley commented on core/server/data/migrations/versions/5.0/2022-04-21-15-44-backfill-mrr-events-for-canceled-subscriptions.js:
```diff
@@ -0,0 +1,66 @@
+            .join('members', 'members_stripe_customers.member_id', '=', 'members.id')
+            .where('members_stripe_customers_subscriptions.cancel_at_period_end', '=', true)
+            .where('members_stripe_customers_subscriptions.mrr', '!=', 0)
+            .where(knex.raw(`(members_paid_subscription_events.type = 'canceled' OR members_paid_subscription_events.type is null)`))
```
any reason not to use the `.where((query) => {...})` syntax for this?

allouis replied:
Not particularly! Do you want me to change it?

matthanley committed:
Switched implementation to use objects to enforce a single event update per subscription

allouis committed:
Split query into two

This is less performant but easier to reason about and fixes the problem
we saw where subscriptions without events were to havign events inserted.

matthanley committed:
Fixed missing field in select

matthanley commented on core/server/data/migrations/versions/5.0/2022-04-21-15-44-backfill-mrr-events-for-canceled-subscriptions.js:
```diff
@@ -0,0 +1,77 @@
+                'subscription_id'
+            )
+            .where('type', '=', 'canceled')
+            .whereIn('subscription_id', canceledSubscriptions.map(x => x.id))
```
Is this safe given that we don't know how many entries we might have?

allouis replied:
As long as we haven't got more than 32k we should be good!

allouis committed:
Reordered migration

allouis commented:
closed in favour of https://github.com/TryGhost/Ghost/pull/14716
